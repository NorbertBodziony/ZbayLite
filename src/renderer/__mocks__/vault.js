export default () => ({
  create: jest.fn(async () => null),
  unlock: jest.fn(async () => null),
  exists: jest.fn(() => false)
})
