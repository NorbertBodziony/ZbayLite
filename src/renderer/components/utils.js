export const getZbayAddress = (zcashAddress) => `zbay.io/uri/${zcashAddress}`

export default {
  getZbayAddress
}
