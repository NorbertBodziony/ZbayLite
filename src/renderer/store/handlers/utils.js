import Immutable from 'immutable'

import { notifierAction } from '../../components/ui/DismissSnackbarAction'

export const typePending = name => `${name}_PENDING`
export const typeFulfilled = name => `${name}_FULFILLED`
export const typeRejected = name => `${name}_REJECTED`

export const errorNotification = ({ message, options }) => ({
  message,
  options: {
    persist: false,
    variant: 'error',
    action: notifierAction,
    ...options
  }
})

export const successNotification = ({ message, options }) => ({
  message,
  options: {
    variant: 'success',
    ...options
  }
})

export const LoaderState = Immutable.Record({
  loading: false,
  message: ''
})

export default {
  typePending,
  typeFulfilled,
  typeRejected
}
