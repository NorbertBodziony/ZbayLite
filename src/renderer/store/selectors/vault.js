import { createSelector } from 'reselect'

const store = s => s

const vault = createSelector(store, state => state.get('vault'))

const exists = createSelector(vault, v => v.exists)
const creating = createSelector(vault, v => v.creating)
// TODO: 14/05 this should be checked on vault, not in redux
const locked = createSelector(vault, v => v.locked)
const unlocking = createSelector(vault, v => v.unlocking)
const error = createSelector(vault, v => v.error)

export default {
  exists,
  creating,
  unlocking,
  error,
  locked
}
