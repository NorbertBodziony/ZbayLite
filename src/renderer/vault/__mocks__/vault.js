import channelsFactory from '../channels'

const workspace = jest.mock()
workspace.save = jest.fn()

export const mock = {
  workspace,
  setArchive: (archive) => {
    workspace.archive = archive
  }
}

export default class Vault {
  constructor (sourceCredentials, archiveCredentials) {
    this.channels = channelsFactory(this)
    this.withWorkspace = async (cb) => cb(workspace)
    this.lock = jest.fn()
  }
}
