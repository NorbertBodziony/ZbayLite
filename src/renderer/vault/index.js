import fs from 'fs'
import path from 'path'

import { remote } from 'electron'

import { passwordToSecureStrings } from './marshalling'
import Vault from './vault'

const getVaultPath = () => path.join(remote.app.getPath('userData'), 'vault.bcup')

let _vault = null

export const exists = () => {
  const userDir = remote.app.getPath('userData')
  const vaultPath = path.join(userDir, 'vault.bcup')
  return fs.existsSync(vaultPath)
}

export const create = async ({ masterPassword }) => {
  if (_vault !== null) {
    throw Error('Archive already initialized.')
  }
  const datasourceObj = {
    type: 'ipc',
    path: getVaultPath()
  }
  const [sourceCredentialsStr, passwordCredentialsStr] = await passwordToSecureStrings({
    masterPassword,
    datasourceObj
  })
  _vault = new Vault(sourceCredentialsStr, passwordCredentialsStr)
}

export const unlock = async ({ masterPassword, createSource = false }) => {
  if (!_vault) {
    await create({ masterPassword })
  }
  try {
    await _vault.unlock(masterPassword, createSource)
  } catch (err) {
    _vault = null
    throw err
  }
}

export const lock = async () => {
  if (!_vault) {
    throw Error('Archive not initialized.')
  }
  return _vault.lock()
}

const _entryToIdentity = entry => {
  const entryObj = entry.toObject()
  return {
    id: entryObj.id,
    name: entryObj.properties.name,
    address: entryObj.properties.address
  }
}
export const createIdentity = async ({ name, address }) => {
  if (!_vault) {
    throw Error('Archive not initialized.')
  }
  let entry = null
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    entry = identitiesGroup.createEntry(name)
      .setProperty('name', name)
      .setProperty('address', address)
    workspace.save()
  })
  return _entryToIdentity(entry)
}

export const listIdentities = async () => {
  if (!_vault) {
    throw Error('Archive not initialized.')
  }
  let identities = []
  await _vault.withWorkspace(workspace => {
    const [identitiesGroup] = workspace.archive.findGroupsByTitle('Identities')
    identities = identitiesGroup.getEntries()
  })
  return identities.map(_entryToIdentity)
}

export const remove = async () => {
  if (_vault) {
    await lock()
    _vault = null
  }
}

export default ({
  identity: {
    createIdentity,
    listIdentities
  },
  remove,
  exists,
  create,
  unlock,
  lock
})
