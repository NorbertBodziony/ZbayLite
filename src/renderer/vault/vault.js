import ChannelQueue from '@buttercup/channel-queue'

import './datasource' // register datasource

import {
  credentialsFromSecureStrings,
  credentialsToSecureStrings,
  credentialsToWorkspace
} from './marshalling'
import { Credentials } from '../vendor/buttercup'
import channelsFactory from './channels'
import contactsFactory from './contacts'
import offersFactory from './offers'

export default class Vault {
  constructor (sourceCredentials, archiveCredentials) {
    if (!Credentials.isSecureString(sourceCredentials)) {
      throw new Error('Failed constructing vault: Source credentials not in encrypted form')
    }
    if (!Credentials.isSecureString(archiveCredentials)) {
      throw new Error('Failed constructing vault: Archive credentials not in encrypted form')
    }
    this._stateQueue = new ChannelQueue()
    this._workspace = null
    this._sourceCredentials = sourceCredentials
    this._archiveCredentials = archiveCredentials
    this.channels = channelsFactory(this)
    this.contacts = contactsFactory(this)
    this.offers = offersFactory(this)
  }

  _enqueueStateChange (cb) {
    return this._stateQueue.channel('state').enqueue(cb)
  }

  locked () {
    return this._workspace === null
  }

  unlock (masterPassword, createSource = false) {
    return this._enqueueStateChange(async () => {
      const [sourceCredentials, archiveCredentials] = await credentialsFromSecureStrings({
        sourceCredentials: this._sourceCredentials,
        archiveCredentials: this._archiveCredentials,
        masterPassword
      })
      const workspace = await credentialsToWorkspace({
        sourceCredentials,
        archiveCredentials,
        createSource
      })
      this._workspace = workspace
      this._sourceCredentials = sourceCredentials
      this._archiveCredentials = archiveCredentials
    })
  }

  lock () {
    return this._enqueueStateChange(async () => {
      const [secSourceCredentials, secArchiveCredentials] = await credentialsToSecureStrings({
        sourceCredentials: this._sourceCredentials,
        archiveCredentials: this._archiveCredentials

      })
      await this._workspace.save()
      this._workspace = null
      this._archiveCredentials = secArchiveCredentials
      this._sourceCredentials = secSourceCredentials
    })
  }

  withWorkspace (cb) {
    return this._enqueueStateChange(async () => cb(this._workspace))
  }
}
