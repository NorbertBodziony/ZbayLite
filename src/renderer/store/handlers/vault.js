import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'

import { typeFulfilled, typeRejected, typePending, errorNotification } from './utils'
import nodeSelectors from '../selectors/node'
import identityHandlers from './identity'
import vaultHandlers from './vault'
import notificationsHandlers from './notifications'
import appHandlers from './app'

import vault from '../../vault'

export const VaultState = Immutable.Record({
  exists: false,
  creating: false,
  unlocking: false,
  creatingIdentity: false,
  locked: true,
  error: ''
}, 'VaultState')

export const initialState = VaultState()

export const actionTypes = {
  CREATE: 'CREATE_VAULT',
  UPDATE_IDENTITY_SIGNER_KEYS: 'UPDATE_IDENTITY_SIGNER_KEYS',
  SET_STATUS: 'SET_VAULT_STATUS',
  UNLOCK: 'UNLOCK_VAULT',
  CREATE_IDENTITY: 'CREATE_VAULT_IDENTITY',
  CLEAR_ERROR: 'CLEAR_VAULT_ERROR'
}

const createVault = createAction(actionTypes.CREATE, vault.create)
const unlockVault = createAction(
  actionTypes.UNLOCK,
  vault.unlock,
  ({ ignoreError = false }) => ({ ignoreError })
)
const createIdentity = createAction(actionTypes.CREATE_IDENTITY, vault.identity.createIdentity)
const updateIdentitySignerKeys = createAction(actionTypes.UPDATE_IDENTITY_SIGNER_KEYS, vault.identity.updateIdentitySignerKeys)
const clearError = createAction(actionTypes.CLEAR_ERROR)
const setVaultStatus = createAction(actionTypes.SET_STATUS)

export const actions = {
  createIdentity,
  updateIdentitySignerKeys,
  createVault,
  unlockVault,
  setVaultStatus,
  clearError
}

const loadVaultStatus = () => (dispatch, getState) => {
  const network = nodeSelectors.network(getState())
  return dispatch(setVaultStatus(vault.exists(network)))
}

const createVaultEpic = ({ name, password }, formActions) => async (dispatch, getState) => {
  const network = nodeSelectors.network(getState())
  try {
    await dispatch(appHandlers.actions.setNewUser(true))
    await dispatch(createVault({ masterPassword: password, network }))
    await dispatch(actions.unlockVault({
      masterPassword: password,
      createSource: true,
      network
    }))
    const identity = await dispatch(identityHandlers.epics.createIdentity({ name }))
    await dispatch(identityHandlers.epics.setIdentity(identity))
    await dispatch(setVaultStatus(true))
  } catch (error) {
    dispatch(notificationsHandlers.actions.enqueueSnackbar(
      errorNotification({ message: `Failed to create vault: ${error.message}` })
    ))
  }
  formActions.setSubmitting(false)
}

const unlockVaultEpic = ({ password: masterPassword }, formActions) => async (dispatch, getState) => {
  const state = getState()
  const network = nodeSelectors.network(state)
  try {
    await dispatch(vaultHandlers.actions.unlockVault({ masterPassword, network, ignoreError: true }))
    const [identity] = await vault.identity.listIdentities()
    await dispatch(identityHandlers.epics.setIdentity(identity))
  } catch (error) {
    if (error.message.includes('Authentication failed')) {
      dispatch(notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({ message: 'Incorrect password' }, { persist: false })
      ))
    } else {
      throw error
    }
  }
  formActions.setSubmitting(false)
}

export const epics = {
  loadVaultStatus,
  createVault: createVaultEpic,
  unlockVault: unlockVaultEpic
}

export const reducer = handleActions({
  [typePending(actionTypes.CREATE)]: state => state.set('creating', true),
  [typeFulfilled(actionTypes.CREATE)]: state => state.merge({
    creating: false,
    exists: true
  }),
  [typeRejected(actionTypes.CREATE)]: (state, { payload: error }) => state.merge({
    creating: false,
    error: error.message
  }),

  [typePending(actionTypes.UNLOCK)]: state => state.set('unlocking', true),
  [typeFulfilled(actionTypes.UNLOCK)]: (state, payload) => state.merge({
    unlocking: false,
    locked: false
  }),
  [typeRejected(actionTypes.UNLOCK)]: (state, { payload: error }) => state.merge({
    unlocking: false,
    locked: true,
    error: error.message
  }),

  [typePending(actionTypes.CREATE_IDENTITY)]: state => state.set('creatingIdentity', true),
  [typeFulfilled(actionTypes.CREATE_IDENTITY)]: state => state.set('creatingIdentity', false),
  [typeRejected(actionTypes.CREATE_IDENTITY)]: (state, { payload: error }) => state.merge({
    creatingIdentity: false,
    error: error.message
  }),
  [setVaultStatus]: (state, { payload: exists }) => state.set('exists', exists),
  [clearError]: state => state.delete('error')
}, initialState)

export default {
  actions,
  epics,
  reducer
}
