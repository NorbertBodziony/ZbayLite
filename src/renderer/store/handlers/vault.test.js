/* eslint import/first: 0 */
jest.mock('../../vault')

import { actions, initialState, actionTypes } from './vault'
import { typePending } from './utils'
import create from '../create'
import vault from '../../vault'

describe('vault reducer', () => {
  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    store = create()
  })

  const assertStoreState = () => expect(store.getState().get('vault')).toMatchSnapshot()

  it('is initialized with vault existence information', async () => {
    expect(initialState.exists).toBeDefined()
    expect(initialState.exists).toEqual(vault.exists())
  })

  it('handles createVault', async () => {
    await store.dispatch(actions.createVault())
    assertStoreState()
  })

  it('handles createVault with loading state', async () => {
    store.dispatch({ type: typePending(actionTypes.CREATE) })
    assertStoreState()
  })

  it('handles createVault error', async () => {
    vault.create.mockImplementationOnce(async () => { throw Error('This is a test error') })
    try {
      await store.dispatch(actions.createVault())
    } catch (err) {}
    assertStoreState()
  })

  it('handles unlockVault', async () => {
    await store.dispatch(actions.createVault())
    await store.dispatch(actions.unlockVault())
    assertStoreState()
  })

  it('handles unlockVault with unlocking state', async () => {
    await store.dispatch(actions.createVault())
    store.dispatch({ type: typePending(actionTypes.UNLOCK) })
    assertStoreState()
  })

  it('handles unlockVault error', async () => {
    await store.dispatch(actions.createVault())
    vault.unlock.mockImplementationOnce(async () => { throw Error('This is a test error') })
    try {
      await store.dispatch(actions.unlockVault())
    } catch (err) {}
    assertStoreState()
  })

  it('handles createIdentity', async () => {
    const identityObj = { name: 'Saturn', addres: 'testaddress' }
    const createdIdentity = {
      id: 'test id',
      ...identityObj
    }
    vault.identity.createIdentity.mockImplementation(async () => createdIdentity)
    const result = await store.dispatch(
      actions.createIdentity(identityObj)
    )
    expect(vault.identity.createIdentity).toHaveBeenCalledWith(identityObj)
    expect(result.value).toEqual(createdIdentity)
    assertStoreState()
  })

  it('handles pending createIdentity', async () => {
    store.dispatch({ type: typePending(actionTypes.CREATE_IDENTITY) })
    assertStoreState()
  })

  it('handles createIdentity error', async () => {
    vault.identity.createIdentity.mockImplementationOnce(async () => { throw Error('This is a test error') })
    try {
      await store.dispatch(actions.createIdentity({ name: 'test', address: 'testaddress' }))
    } catch (err) {}
    assertStoreState()
  })

  it('handles clear error', async () => {
    const errorMsg = 'This is a test error'
    vault.unlock.mockImplementationOnce(async () => { throw Error(errorMsg) })
    try {
      await store.dispatch(actions.unlockVault())
    } catch (err) {}

    const vaultStore = store.getState().get('vault')
    expect(vaultStore.error).toEqual(errorMsg)

    store.dispatch(actions.clearError())
    assertStoreState()
  })
})
