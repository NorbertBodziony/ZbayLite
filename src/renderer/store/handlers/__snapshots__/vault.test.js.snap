// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vault reducer handles clear error 1`] = `
Immutable.VaultState {
  "exists": false,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": true,
  "error": "",
}
`;

exports[`vault reducer handles createIdentity 1`] = `
Immutable.VaultState {
  "exists": false,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": true,
  "error": "",
}
`;

exports[`vault reducer handles createIdentity error 1`] = `
Immutable.VaultState {
  "exists": false,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": true,
  "error": "This is a test error",
}
`;

exports[`vault reducer handles createVault 1`] = `
Immutable.VaultState {
  "exists": true,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": true,
  "error": "",
}
`;

exports[`vault reducer handles createVault error 1`] = `
Immutable.VaultState {
  "exists": false,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": true,
  "error": "This is a test error",
}
`;

exports[`vault reducer handles createVault with loading state 1`] = `
Immutable.VaultState {
  "exists": false,
  "creating": true,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": true,
  "error": "",
}
`;

exports[`vault reducer handles pending createIdentity 1`] = `
Immutable.VaultState {
  "exists": false,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": true,
  "locked": true,
  "error": "",
}
`;

exports[`vault reducer handles unlockVault 1`] = `
Immutable.VaultState {
  "exists": true,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": false,
  "error": "",
}
`;

exports[`vault reducer handles unlockVault error 1`] = `
Immutable.VaultState {
  "exists": true,
  "creating": false,
  "unlocking": false,
  "creatingIdentity": false,
  "locked": true,
  "error": "This is a test error",
}
`;

exports[`vault reducer handles unlockVault with unlocking state 1`] = `
Immutable.VaultState {
  "exists": true,
  "creating": false,
  "unlocking": true,
  "creatingIdentity": false,
  "locked": true,
  "error": "",
}
`;
